packages/jupyter-ai/jupyter_ai/document_loaders/directory.py

```diff
@@ -119,17 +132,29 @@ def collect_filepaths(path, all_files: bool):
     # Check if the path points to a single file
     if os.path.isfile(path):
         filepaths = [Path(path)]
+    elif os.path.isdir(path):
+        filepaths = walk_directory(path, all_files)
     else:
         filepaths = []
-        for dir, subdirs, filenames in os.walk(path):
-            # Filter out hidden filenames, hidden directories, and excluded directories,
-            # unless "all files" are requested
-            if not all_files:
-                subdirs[:] = [
-                    d for d in subdirs if not (d[0] == "." or d in EXCLUDE_DIRS)
-                ]
-                filenames = [f for f in filenames if not f[0] == "."]
-            filepaths.extend([Path(dir) / filename for filename in filenames])
+        directories = []
+        for glob_path in iglob(str(path), include_hidden=all_files, recursive=True):
+            if os.path.isfile(glob_path):
+                filepaths.append(Path(glob_path))
+                continue
+            # only keep directories if no files
+            elif len(filepaths) == 0:
+                directories.append(glob_path)
+        # walk the directories if only directories are returned from iglob
+        if not filepaths and directories:
+            filepaths = list(
+                {
+                    Path(path)
+                    for dir_paths in map(
+                        lambda p: walk_directory(p, all_files), directories
+                    )
+                    for path in dir_paths
+                }
+            )
```

Comment:
I think that having conditional behavior on whether the glob matches only directories or not can be very confusing to less technical users. For simplicity, it would be better to only allow globs to match files, while completely ignoring directories matched by the glob.

Note that this would still support matching files under multiple directories in a single argument. Consider the example you provided earlier:

```
folder1/
| file1
| file2
folder2/
| file3
| file4
unwanted_folder/
| file_a
| file_b
test.py
```

You should be able to still query all files under `folder1/` and `folder2/` by calling:

```
/learn folder*/*
```

If you want to do this recursively for subdirs for `folder1/` and `folder2/`, you should still be able to do this by calling:

```
/learn folder*/**/*
```

Can you try deleting the code that highlighted here and try the above two cases?

Reply:
Those work! I deleted the code and pushed up the changes.